jetty-ee10/jetty-ee10-webapp/src/main/java/org/eclipse/jetty/ee10/webapp/WebAppClassLoader.java

```diff
@@ -299,32 +308,35 @@ public void addJars(Resource lib)
     {
         if (lib.exists() && lib.isDirectory())
         {
-            List<String> entries = lib.list();
-            if (entries != null)
+            Path dir = lib.getPath();
+
+            try (Stream<Path> streamEntries = Files.list(dir))
             {
-                entries.sort(Comparator.naturalOrder());
+                List<Path> jars = streamEntries
+                    .filter(Files::isRegularFile)
+                    .filter(this::isFileSupported)
+                    .sorted(Comparator.naturalOrder())
+                    .toList();
 
-                for (String entry : entries)
+                for (Path jar: jars)
                 {
                     try
                     {
-                        Resource resource = lib.resolve(entry);
                         if (LOG.isDebugEnabled())
-                            LOG.debug("addJar - {}", resource);
-                        String fnlc = resource.getName().toLowerCase(Locale.ENGLISH);
-                        // don't check if this is a directory (prevents use of symlinks), see Bug 353165
-                        if (isFileSupported(fnlc))
-                        {
-                            String jar = URIUtil.encodeSpecific(resource.toString(), ",;");
-                            addClassPath(jar);
-                        }
+                            LOG.debug("addJar - {}", jar);
+                        URI jarUri = Resource.toJarFileUri(jar.toUri());
+                        addClassPath(jarUri.toASCIIString());
```

Comment:
Another `toASCIIString` that worries me w.r.t files with non-ascii characters in their names.

Reply:
Using `toASCIIString` doesn't strip characters, it ensure that the entire string is ASCII characters, even if the original input string contains unicode.